src/unit.rs

```diff
@@ -2153,6 +2174,160 @@ impl<'input, 'abbrev, 'unit, Endian> EntriesCursor<'input, 'abbrev, 'unit, Endia
             }
         }
     }
+
+    /// Return a tree view of the entries that have the current entry as the root.
+    pub fn tree(self) -> EntriesTree<'input, 'abbrev, 'unit, Endian> {
+        EntriesTree::new(self)
+    }
+}
+
+/// The state information for a tree view of the Debugging Information Entries.
+///
+/// The `EntriesTree` can be used to recursively iterate through the DIE
+/// tree, following the parent/child relationships. It maintains a single
+/// `EntriesCursor` that is used to parse the entries, allowing it to avoid
+/// any duplicate parsing of entries.
+#[derive(Clone, Debug)]
+pub struct EntriesTree<'input, 'abbrev, 'unit, Endian>
+    where 'input: 'unit,
+          Endian: Endianity + 'unit
+{
+    cursor: EntriesCursor<'input, 'abbrev, 'unit, Endian>,
+    // The depth of the entry that cursor::next_sibling() will return.
+    depth: isize,
+}
+
+impl<'input, 'abbrev, 'unit, Endian> EntriesTree<'input, 'abbrev, 'unit, Endian>
+    where Endian: Endianity
+{
+    fn new(cursor: EntriesCursor<'input, 'abbrev, 'unit, Endian>) -> Self {
+        EntriesTree {
+            cursor: cursor,
+            depth: 0,
+        }
+    }
+
+    /// Returns an iterator for the entries that are children of the current entry.
+    ///
+    /// Generally, this function should only be called when the `EntriesTree`
+    /// is first created.  This will give an iterator for the children of the
+    /// root entry.  The result of subsequent calls to this function is unspecified.
```

Comment:
> We should just make it take self instead of &mut self then, IMO. This would prevent subsequent calls from being possible. Any reason why you didn't pursue this?

We would need to return both a new `self` and the iterator, and the iterator needs to be a borrow of the new `self`, which I don't think is possible.

> Or maybe assert! against calling it more than once, if we need all the EntryTreeIters to share a single &mut EntriesTree reference?

We would need add a new flag to assert against. If we were going to take that approach, then I would prefer to instead make it possible to call `iter` more than once (I think this would simply need to store the starting offset of the cursor).